Font Face: Get name from "fontFamily" setting, not "name".

hellofromtonya opened the pull request:
## What?

Get the font-family name from the "fontFamily" field, instead of  the"name" field.

## Why?
To avoid a back-compatibility (BC) break for themes that do not have the `"name"` setting defined.

WP Core also does not require the `"name"` setting. Using it exclusively is a BC break in Core. 

It's not clearly documented that the `"name"` setting is required. Thus, it's been optional.

## How?

Use the font-family's `fontFamily` field instead of  its `name` field.

As the `fontFamily` field could be a comma-separated list of font-families, it parses the list to extract the first entry, using the same approach in the Fonts API.

### Why `fontFamily` setting?

`WP_Theme_JSON` schema requires the `fontFamily` setting in each of the `typography.fontFamilies`. 

## Testing Instructions

1. Use WordPress 6.3, not Core`trunk`. (Why? `trunk` has Font Face in it, so the Font Face files in Gutenberg will not load into memory.) 
2. Activate the Twenty-Twenty-Three (TT3) theme.
3. Open the TT3 `theme.json` file in your favorite editor and modify it as follows:
    * Scroll to line 152 and delete the `"name": "DM Sans",` line.
    * Save.
4. In your browser, open its dev tools and in the HTML inspector, search for and expand the `wp-fonts-local` style element.

Expected:
* Without this PR fix, DM Sans should not be included in the `@font-face` styles ❌ 
* With this PR fix, DM Sans should be included in the `@font-face` styles ✅

hellofromtonya committed:
Gets name from fontFamily as fallback check.

If the "name" setting does not exist, then get it from the "fontFamily" setting.

hellofromtonya committed:
Move tests into existing test.

To avoid reflection issues on different PHP versions and
potential contributor confusion, merging the new tests into
the existing test.

pbking commented:
If the name has ' symbols those are included.  We should also strip out any ' or "

matiasbenedetto commented:
:wave:

I think we should not default to using the name of a font family as the default value for the CSS font faces printed into the site. If the `name` is not exactly the same as one of the values of `fontFamily` that breaks. Let's say that a theme adds a font family like this:

```json
{
	"fontFace": [
		{
			"fontFamily": "Cardo",
			"fontStyle": "normal",
			"fontWeight": "700",
			"src": [
				"file:./assets/fonts/cardo_normal_700.woff2"
			]
		}
	],
	"fontFamily": "Cardo",
	"name": "Heading",
	"slug": "heading"
}
```
The font face css definition in this case is:

```css
@font-face {
  font-family: Heading;
  font-style: normal;
  font-weight: 700;
  font-display: fallback;
  src: url("http://localhost/wp1/wp-content/themes/twentytwentyfour/assets/fonts/cardo_normal_700.woff2")
    format("woff2");
}
```
And the css preset is:
```
--wp--preset--font-family--heading: Cardo;
```

So, the font is not rendering as expected.

matiasbenedetto commented:
I think the backup of using the first value of the family `fontFamily` property seems to be working better and seems much more difficult to break. But adds complexity that may be avoided. What do you think of using the font face `fontFamily` property to render the font face CSS definitions (`@font-face `) ?

So following the previous example we could have something like the following:


theme.json font family definition

```json
{
	"fontFace": [
		{
			"fontFamily": "Cardo",
			"fontStyle": "normal",
			"fontWeight": "700",
			"src": [
				"file:./assets/fonts/cardo_normal_700.woff2"
			]
		}
	],
	"fontFamily": "Cardo, system-ui",
	"name": "Heading",
	"slug": "heading"
}
```
Output css for font face
```css
@font-face {
  font-family: Cardo;
  font-style: normal;
  font-weight: 700;
  font-display: fallback;
  src: url("http://localhost/wp1/wp-content/themes/twentytwentyfour/assets/fonts/cardo_normal_700.woff2")
    format("woff2");
}
```
CSS preset value:

```css
--wp--preset--font-family--heading: Cardo;
```

pbking commented:
> What do you think of using the font face fontFamily property to render the font face CSS definitions (@font-face ) ?

That collection isn't required right?  System fonts would not.  However there will always be a "first item in the list" for fontFamily.

I think it would be great to use that value for the `@font-face` value though.  There won't be one of those for system fonts anyway.  It's just not the best backup for 'name' but perhaps the best source for `@font-face`.

pbking commented:
Before this change the 'Name' rendered in the editor was derived some other way when the `name:` attribute was missing and the fontFamily was a list the whole list wasn't shown.  With this change the whole value of fontFamily is displayed in the editor UI
<img width="136" alt="image" src="https://github.com/WordPress/gutenberg/assets/146530/b1989b77-62d6-4b13-9709-b072e8bf732a">

hellofromtonya commented:
>I think the backup of using the first value of the family fontFamily property seems to be working better and seems much more difficult to break. But adds complexity that may be avoided. What do you think of using the font face fontFamily property to render the font face CSS definitions (@font-face ) ?

@matiasbenedetto Hmm what if the `name` is different from the `fontFamily`? Why would it be?

If the `name` exists, I think that's being used to populate the Typography pickers in the editors. Having a different `name` would be confusing to users. But an extender may have reasons to make the 2 setting values different.

So to protect the styles, I think you're right - it should use the required `fontFamily` and not the `name` setting.

matiasbenedetto commented:
> it should use the required fontFamily and not the name setting.

Yep, just to reiterate I would prefer to use the font face fontFamily property instead of the family `fontFamily` one. Both are required fields on `theme.json` schema, but I think it would be better to use the font face one because it shouldn't need extra processing (spliting by comma) and seems like a good place to indicate that parameter explicitly.

hellofromtonya commented:
>Before this change the 'Name' rendered in the editor was derived some other way when the `name:` attribute was missing and the `fontFamily` was a list the whole list wasn't shown. With this change the whole value of fontFamily is displayed in the editor UI

@pbking Font Face is read-only of Theme JSON. It's not involved in what gets populated in the typography pickers.

Testing it with only Core's `trunk` and with WP 6.3.1 (i.e. not with this PR or the Gutenberg plugin), I see the same problem:

<img width="329" alt="Screenshot 2023-09-19 at 2 56 12 PM" src="https://github.com/WordPress/gutenberg/assets/7284611/eed91ec4-dbf1-496d-9404-5d799b03e6e5">

You found a bug. Want to open a new issue for it?

hellofromtonya commented:
>Yep, just to reiterate I would prefer to use the font face fontFamily property instead of the family fontFamily one. Both are required fields on theme.json schema, but I think it would be better to use the font face one because it shouldn't need extra processing (spliting by comma) and seems like a good place to indicate that parameter explicitly.

@matiasbenedetto What happens though if an extender has copied and pasted the family `fontFamily` value? It could also have a comma-separated list of font families.

hellofromtonya commented:
>That collection isn't required right? System fonts would not. However there will always be a "first item in the list" for fontFamily.

@pbking The System fonts though do not have `fontFace` variations, right? The first check (before getting the font-family name) skips the font if there's no `fontFace` declared.

hellofromtonya committed:
Do not use "name".

Do not the "name" settting. Instead only use the required "fontFamily" setting.

matiasbenedetto commented:
>  @matiasbenedetto What happens though if an extender has copied and pasted the family fontFamily value? It could also have a comma-separated list of font families.

It seems less likely, but in that case, we can maintain the coma-splitting safety. Apart from that, It would be nice to document each field properly so extenders know what is expected in each field.

hellofromtonya commented:
>It seems less likely, but in that case, we can maintain the coma-splitting safety.

I think the comma-splitting is needed, just in case.

>Apart from that, It would be nice to document each field properly so extenders know what is expected in each field.

Do you mean each field in the `theme.json` file?

matiasbenedetto commented:
> Do you mean each field in the `theme.json` file?

Yes, but that's not part of the scope of this PR. Just a nice thing to have in the future.

hellofromtonya commented:
>Yes, but that's not part of the scope of this PR. Just a nice thing to have in the future.

100% I agree with you! That would be helpful for everyone (me included).

ironprogrammer commented:
## Test Report

### Environment
- OS: macOS 13.5.2
- Browser: Safari 16.6
- Server: nginx/1.25.2
- PHP: 7.4.33
- WordPress: 6.3.1
- Theme: twentytwentythree v1.2
- Active Plugins:
  - gutenberg v16.6.0

### Actual Results
- ✅ DM Sans is included in `#wp-fonts-local`.
- ✅ The first parsed value in the `fontFamily` prop is used in the rendered `font-family` CSS. E.g. in `theme.json` using `"fontFamily": "sans-serif, \"Inter\""` (reversing the order) results in `@font-face` having `font-family:sans-serif`.
- ✅ The value provided in the `name` prop can be arbitrary, and does not affect the rendered CSS font family name.

### Additional Notes
Observations that can be followed-up separately if needed:
- The value provided in the `name` prop is used to identify the fonts in *Styles > Typography* (DM Sans in Figure 1).
- Confirmed that [as noted earlier](https://github.com/WordPress/gutenberg/pull/54615#issuecomment-1726351370), when no `name` prop is present, the entire value of `fontFamily` is used to label the font in the UI (Inter in Figure 1).
- If neither a `fontFace` or `fontFamily` prop are provided, the font still shows up in the Fonts list, though has no corresponding rendered `@font-face`.

### Supplemental Artifacts
*Figure 1: Displayed picker font names versus rendered font faces.*
<img width="666" alt="font names versus font faces" src="https://github.com/WordPress/gutenberg/assets/824344/be665c9c-29ed-4512-a342-4ad13b45e0f4">

ironprogrammer commented on lib/compat/wordpress-6.4/fonts/font-face/class-wp-font-face-resolver.php:
```diff
@@ -56,30 +56,54 @@ private static function parse_settings( array $settings ) {
+			$font_family = explode( ',', $font_family )[0];
+		}
+
+		return $font_family;
```
```suggestion
		return trim( $font_family, "\"'" );
```
Should this be stripped of wrapping quotes before returning? Some of the testing today included various use cases where this might be a factor. This would also address the unit test failure happening [here](https://github.com/WordPress/gutenberg/actions/runs/6240465587/job/16940632328?pr=54615#step:14:130). (For reference, [`phpunit/data/themedir1/fonts-block-theme/theme.json`](https://github.com/WordPress/gutenberg/blob/74e1db8c535115eecaf8cfc00db8eef3a3d8b9cd/phpunit/data/themedir1/fonts-block-theme/theme.json) uses escaped quotes in both of the `fontFamily` values in question, which with this PR are now the values returned in the test.)

ironprogrammer commented on phpunit/tests/fonts/font-face/wpFontFaceResolver/getFontsFromThemeJson.php:
```diff
@@ -101,4 +101,108 @@ public function data_should_replace_src_file_placeholder() {
+				'fonts'         => array(
+					array(
+						'fontFamily' => 'DM Sans',
+						'name'       => 'DM Sans',
```
```suggestion
						'name'       => 'DM Sans Family',
```
How about making sure the `name` value differs from `fontFamily` to validate that the former is not being used?

ironprogrammer commented on phpunit/tests/fonts/font-face/wpFontFaceResolver/getFontsFromThemeJson.php:
```diff
@@ -101,4 +101,108 @@ public function data_should_replace_src_file_placeholder() {
+			'fontFamily comma-separated list' => array(
+				'fonts'         => array(
+					array(
+						'fontFamily' => 'DM Sans, sans-serif',
```
```suggestion
						'fontFamily' => '"DM Sans", sans-serif',
```
If you agree with trimming wrapping quotes in `maybe_parse_name_from_comma_separated_list()`, this would help validate that change as well. TT3 uses double-quotes around font names, except for IBM Plex Mono (single-quotes) and System Font (not quoted).

ironprogrammer reviewed (changes requested):
Thanks for this PR, @hellofromtonya! I think this is the right direction (I actually love being able to indicate arbitrary font names in the UI). Here are a few things to consider as this wraps up.

matiasbenedetto committed:
trim quotes on font family name

Co-authored-by: Brian Alexander <824344+ironprogrammer@users.noreply.github.com>

matiasbenedetto committed:
making sure the name value differs from fontFamily to validate that the former is not being used

Co-authored-by: Brian Alexander <824344+ironprogrammer@users.noreply.github.com>

matiasbenedetto committed:
validate trimming changes

Co-authored-by: Brian Alexander <824344+ironprogrammer@users.noreply.github.com>

matiasbenedetto committed:
Merge branch 'trunk' into fix/font-face/backup-font-family-name-check

ironprogrammer committed:
Merge branch 'trunk' into fix/font-face/backup-font-family-name-check

ironprogrammer reviewed (approved):
Refreshed with `trunk` and LGTM 👍🏻

ironprogrammer committed:
Remove test temporarily

Because `WP_Font_Face_Resolver` already exists in Core release code, this update cannot be tested in CI without migrating the code into a temporary `_Gutenberg` resolver, which will complicate syncing back to Core.

Test will be added to a follow-up PR, to merge once this update is synced to Core.

Note that this test does succeed when run locally against WP `trunk` (currently 6.4-alpha).

mikachan commented:
I just cherry-picked this PR to the release/16.7 branch to get it included in the next release: 5cde445574

hellofromtonya commented:
PR 5298 https://github.com/WordPress/wordpress-develop/pull/5298 ports the Resolver changes and tests to Core. Once committed, the tests can be (re) added into the Gutenberg repo for the CI jobs running against Core's `trunk` (those that were previously failing).

bph commented:
Solves: 

- #54605
